### DISCUSSION TITLE
How mqtt connection is distributed to clustered(on k8s) emqx nodes?

### DISCUSSION THREAD

**Discussion by gomgomshrimp** [emqx/emqx#4532] [Q&A] [answered]
I'm running emqx cluster on k8s and I checked mqtt connection is distributed to nodes. 

I read that emqx cluster is usually deployed behind LB... but I'm not built any LB for emqx cluster on k8s.  

Then how mqtt connection is distributed to nodes? Is there its own LB when building emqx cluster on k8s?

**Comment by hjianbo** [comment]
It depends only on which node the client is connected to, and on the front Proxy's strategy for distributing TCP connections.

For k8s, `Ingress` may be useful for you https://kubernetes.io/docs/concepts/services-networking/ingress/

**Comment by gomgomshrimp** [comment]
@HJianBo Thanks to reply.
But there are some misunderstood for my question. I'm not asking about connection from external to internal...
 
I already built the emqx cluster on EKS and it works well. There are three emqx nodes for clustering, and mqtt connections are distributed well, but I didn't set any Load Balancer. (Each emqx node uses "headless service" - kubernetes service that has no load balancing)

Please check emqx cluster by using [autocluster based k8s](https://docs.emqx.io/en/broker/v4.3/advanced/cluster.html#node-discovery-and-automatic-clustering). Maybe Kubernetes does it automatically not emqx....

**Comment by hjianbo** [comment]
Sorry @gomgomshrimp I don't know much about EKS, but I'm sure there must be services in the middle that exist for LB functions. Because EMQ X nodes can't do connection distribution by themselves. We formed the cluster by configuring K8S just to find the node names of other emqx nodes.

**Comment by bloodberrys** [comment]
@gomgomshrimp Well, I have the same question to this issue. Did you have any solution already? Anyway, I deployed the emqx using helm charts based on the emqx broker doc here: https://docs.emqx.io/en/broker/v4.3/getting-started/install.html#install-emq-x-in-docker-contain-a-simple-docker-compose-cluster to my EKS, but there was no loadBalancer available on the emqx-service. All tutorial available for fully-managed kubernetes (such as kubeadm, minikube, etc), but not particularly provided for EKS. Then, how to expose this service in EKS?
![image](https://user-images.githubusercontent.com/36856722/114297395-921f3980-9ada-11eb-8dcf-9de200eb89f8.png)

**Comment by bloodberrys** [comment]
I have tried to patch the service to 
`kubectl patch svc my-emqx -p '{"spec": {"type": "LoadBalancer"}}'`
and also edit the service to add an NLB loadbalancer annotations for AWS. The external IP showed up but I still can't pub/sub with the DNS.

**Comment by zmstone** [comment]
If I remember correctly, service type can not be changed without redeployment.

**Comment by zmstone** [comment, marked as answer]
> @HJianBo Thanks to reply.
> But there are some misunderstood for my question. I'm not asking about connection from external to internal...
> 
> I already built the emqx cluster on EKS and it works well. There are three emqx nodes for clustering, and mqtt connections are distributed well, but I didn't set any Load Balancer. (Each emqx node uses "headless service" - kubernetes service that has no load balancing)
> 
> Please check emqx cluster by using [autocluster based k8s](https://docs.emqx.io/en/broker/v4.3/advanced/cluster.html#node-discovery-and-automatic-clustering). Maybe Kubernetes does it automatically not emqx....

Hi. I am not sure if I understand the questions correctly but
1.  EQM X nodes discover each other to form a cluster using k8s APIs.
2. K8s Service publishing (so to reachable from external)

These are two different topics.
I don't think they are related.

I think @bloodberrys is going in the right direction (by adding a load balancer).
ref: https://kubernetes.io/docs/concepts/services-networking/service/#publishing-services-service-types

**Comment by bloodberrys** [comment]
@zmstone exactly. 
This is my attempt:
0. Adding tcp port 1883 to my AWS security group.
1. `kubectl edit svc/my-emqx` and adding `service.beta.kubernetes.io/aws-load-balancer-type: "nlb"` under annotations.
2. exit
3. `kubectl patch svc my-emqx -p '{"spec": {"type": "LoadBalancer"}}'`
4. I got my external IP!!
![image](https://user-images.githubusercontent.com/36856722/114298390-5edfa900-9ae0-11eb-9ead-d9edc6c8bf0b.png)
5. `nslookup` my external IP:
![image](https://user-images.githubusercontent.com/36856722/114298403-7159e280-9ae0-11eb-91fb-e7662432cae9.png)
seems great.

but when I Tried to pub/sub:
![image](https://user-images.githubusercontent.com/36856722/114298441-a36b4480-9ae0-11eb-81c5-a917d4602284.png)
I got `All subscription requests were denied.`. 

What does `All subscription requests were denied.` stand for?

**Comment by zmstone** [comment]
I think clients are not allowed to subscribe `#`.
you can try to subscribe `presence` (as where the producer publishes message to)
or `prefix/#`

**Reply by bloodberrys** [reply]
yes, it works.!!!
You are right. sorry for that. thanks, @zmstone Now it works:
![image](https://user-images.githubusercontent.com/36856722/114299024-6a809f00-9ae3-11eb-926e-d97410ec085f.png)


but why? I usually use this wildcard ('#') on emqx local and it still works properly:
![image](https://user-images.githubusercontent.com/36856722/114299143-f692c680-9ae3-11eb-869c-5fbc80d7b78b.png)

**Reply by zmstone** [reply]
Great.

This is because of the default ACLs, see`acl.conf`
```
{allow, {ipaddr, "127.0.0.1"}, pubsub, ["$SYS/#", "#"]}.
{deny, all, subscribe, ["$SYS/#", {eq, "#"}]}.
```

**Reply by bloodberrys** [reply]
aah I see, is this for security reasons?

**Reply by zmstone** [reply]
Yeah. Subscribing to `#` basically means a DoS attach in many cases.

**Comment by gomgomshrimp** [comment]
@zmstone @bloodberrys
Guys thank you for answer my question.
There are still some differences from the intent of my question. I'm sorry that my english is not good enough...

My question is **"is there something(such as server...)inside emqx cluster for distributing mqtt connections to nodes"**.

This is my emqx dashboard. You can see that connections are distributed to nodes. (emqx-0 : 2 connections, emqx-1: 2 connections, emqx-2: 1 connection)
![스크린샷 2021-04-14 오후 4 30 20](https://user-images.githubusercontent.com/12162614/114672039-ad3eb300-9d3f-11eb-9eb4-da8cfba368ae.png)

I'm not using any Loadbalancing service. There are only service named "iot-service", and it is 'Headless service' (doesn't have loadbalance too). 
![스크린샷 2021-04-14 오후 4 41 05](https://user-images.githubusercontent.com/12162614/114672764-761cd180-9d40-11eb-8f96-e3685aa2319a.png)

All connections use the same internal eks dns address "iot-service.default.svc.cluster.local"... But connections are distributed to nodes as you see.

It is not pretty big issue, but emqx docs says "[LoadBalancer distributes MQTT connections and traffic from devices across the EMQ X clusters](https://docs.emqx.io/en/broker/v4.3/tutorial/deploy.html#load-balancer-lb)".

**Reply by bloodberrys** [reply]
I think you need to try to patch your service to have a Network Load Balancer, and see how it works. NLB is particularly for TCP connections. It will distribute ur tcp connections automatically on the available node target. See here for more explanations: https://docs.aws.amazon.com/elasticloadbalancing/latest/network/introduction.html

**Reply by bloodberrys** [reply]
You could follow the step As per my attempt above to patch it. You can try it, whether it is distributing your expected result or not.

**Reply by zmstone** [reply]
Hi @gomgomshrimp 
So far, we do not have a product/feature for MQTT load balancing
not EMQ X internally, nor externally.

Balancing solutions that I know of are

1. proxies such as nginx (TCP layer)
2. network load balancers (TCP layer)
3. DNS roundrobin
4. client-id hash based balancing in nginx (MQTT layer) [1] (this is quite interesting, but it's not free, and I have not tried it myself)

[1] https://www.nginx.com/blog/nginx-and-iot-adding-protocol-awareness-for-mqtt/

**Reply by bloodberrys** [reply]
@gomgomshrimp I agreed with @zmstone, those (4) points are reasonable. I also have read this paper [scalablemqtt.pdf](https://github.com/emqx/emqx/files/6316960/scalablemqtt.pdf) By using method number 4,
different ClientIDs will be hashed and thoroughly CONNECT to a different message broker.
